design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+[Objective-C](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418746-description)
+have a mechanism to retrieve the string representation of an object. 
+
+The StringRepresentation information category retrieves the string representation of the object in
+question. The property type for StringRepresentation is an xlang_string.
+
+While all information categories are optional, it is highly recommended that all objects support
+StringRepresentation. By default, this method can return the type's name, however it can be
+implemented in whatever manner makes most sense for the underlying object.
+
+#### Type Information
+
+Dynamic languages such as JavaScript and Python often need to be able to discover information
+about an object at runtime. In order to do this, xlang object must have a mechanism to find the
+metadata for arbitrary xlang objects.
+
+The TypeName information category retrieves the fully namespace qualified name of the type in
+question. The property type for TypeName is an xlang_string.
+
+While all information categories are optional, it is highly recommended that all objects support
+TypeName in order to be usable from dynamic languages.
+
+#### Memory Usage
+
+In languages C# and Java, garbage collection happens on background threads on a non-deterministic
+schedule. These garbage collectors typically depend on an omniscient understanding of memory usage.
+The memory used by xlang components is not typically visible to these garbage collectors, leading
+to uninformed decision making regarding collection scheduling. If these systems could retrieve
+the memory usage of arbitrary xlang objects, their scheduling heuristics would be better and the
+system would run more smoothly overall.
+
+The MemoryUsage information category retrieves the memory usage in bytes of the object in question.
+The property type for MemoryUsage is an unsigned 32-bit integer.
+
+## Weak References
+
+One issue faced in COM based systems is circular references - i.e. two objects who each hold a
+pointer to the other. In this case, neither reference count will ever go to zero and neither
+object will ever be cleaned up. To avoid this issue, xlang supports weak references.
+
+A weak reference to an xlang object does not keep it alive. In other words, when all non-weak
+reference count of an object goes to zero, it cleans itself up even if there are outstanding weak
+references to it. As such, weak reference holders must check to see if a given reference is still
+valid via the IWeakReference::Resolve method.
+
+Weak references are retrieved for an arbitrary xlang object by querying for the
+IWeakReferenceSource interface and calling GetWeakReference. If an object does not support weak
+references, it simply returns the no interface error code from query interface.
+
+Note, because IWeakReferenceSource and IWeakReference are ABI interfaces that represents a proxy to
```

Comment:
Another way to put it is that because weak references are not part of the projected surface of any API. Specifically, IXlangObject is only the base interface for projected types and a weak reference is not a projected type.

Reply:
Good suggestion.

Would you say IActivationFactory (as well as custom activation factory interfaces) should inherit from IXlangObject, IUnknown, or some other base interface specific to activation factory interfaces?

Reply:
I don't see any problem with factory interfaces inheriting from IXlangObject, but there is no real need for them to do so, other than that it complicates the IDL/winmd authoring/generation process as you would have to distinguish between factory and non-factory interfaces. I'd rather not create another sub-category of interface.